### PR TITLE
videosource: Support reading HDR metadata from container

### PR DESCRIPTION
**Pull request by kodabb**
Read AVMasteringDisplayMetadata and AVContentLightMetadata from
stream side data when present, and use it to populate VideoProperties
when frame side data is absent.

### CODE REVIEW

File: src/core/videosource.cpp [hunk 1]
```diff
@@ -245,11 +245,41 @@ FFMS_VideoSource::FFMS_VideoSource(const char *SourceFile, FFMS_Index &Index, in
         VP.Stereo3DFlags = 0;
 
         for (int i = 0; i < FormatContext->streams[VideoTrack]->nb_side_data; i++) {
-            if (FormatContext->streams[VideoTrack]->side_data->type == AV_PKT_DATA_STEREO3D) {
-                const AVStereo3D *StereoSideData = (const AVStereo3D *)FormatContext->streams[VideoTrack]->side_data->data;
+            if (FormatContext->streams[VideoTrack]->side_data[i].type == AV_PKT_DATA_STEREO3D) {
```
Comment (dwbuiten) [inline src/core/videosource.cpp]: Is this a second change/fix mixed in (iterating properly), or am I misunderstanding?

File: src/core/videosource.cpp [hunk 2]
```diff
@@ -245,11 +245,41 @@ FFMS_VideoSource::FFMS_VideoSource(const char *SourceFile, FFMS_Index &Index, in
...
+                /* MasteringDisplayMinLuminance can be 0 */
+                VP.HasMasteringDisplayLuminance = !!VP.MasteringDisplayMaxLuminance;
+            } else if (FormatContext->streams[VideoTrack]->side_data[i].type == AV_PKT_DATA_CONTENT_LIGHT_LEVEL) {
+                const AVContentLightMetadata *ContentLightLevel = (const AVContentLightMetadata *)FormatContext->streams[VideoTrack]->side_data[i].data;
+
+                VP.ContentLightLevelMax = ContentLightLevel->MaxCLL;
+                VP.ContentLightLevelAverage = ContentLightLevel->MaxFALL;
+
+                /* Only check for either of them */
+                VP.HasContentLightLevel = !!VP.ContentLightLevelMax || !!VP.ContentLightLevelAverage;
```
Comment (dwbuiten) [inline src/core/videosource.cpp]: Not too familiar with HDR stuff, but is it ever useful to say which?

Comment (kodabb) [reply src/core/videosource.cpp]: in practice it never happens that only one variable is set and the other is not, but other tools like x265 use the same check

`    m_emitCLLSEI = p->maxCLL || p->maxFALL;`

**Comment by kodabb** [comment]
> Are there ever cases where a file may have both frame level and container/stream level mastering metadata?

A user could have frame-level metadata and then they could use mkvmerge/youtube-hdr-tool to fiddle the with the container metadata. Philosophically I prefer to give priority to parameters that are coming from the frames, as they are harder to modify and less likely to be interpreted wrong.

**Comment by dwbuiten** [comment]
OK. is it possible to add a small unit test for each type? There should be some free-to-use samples, I think?

**Comment by kodabb** [comment]
thanks for the reviews and code @dwbuiten!

**Pull request merged**